feat(agent-tool): Add event streaming propagation from AgentTool sub-agents to Runner

sarojrout opened the pull request:
Implements real-time event streaming for AgentTool to propagate sub-agent events, addressing issue #3984.

### Link to Issue or Description of Change

**1. Link to an existing issue (if applicable):**

- Closes: #3984

**2. Or, if no issue exists, describe the change:**

_If applicable, please follow the issue templates to provide as much detail as
possible._

**Problem:**
When a coordinator agent delegates to a sub-agent via `AgentTool`, the sub-agent's execution acts as a "black box". No events are yielded during sub-agent execution, making the frontend appear unresponsive for the duration of sub-agent execution (which can be 20-30 seconds for complex tasks). Only the final result is returned after sub-agent completes, providing no visibility into intermediate steps, tool calls, or responses.

This creates a poor user experience for hierarchical multi-agent systems where users need real-time feedback about sub-agent progress.

**Solution:**
Implemented event streaming propagation for `AgentTool` to yield events from sub-agents in real-time to the parent Runner. This enables:
1. **Real-time visibility**: Sub-agent events (model responses, tool calls, tool responses) are streamed immediately as they are generated
2. **Progressive feedback**: Frontend receives events progressively, not all at once at the end
3. **Better UX**: Users can see intermediate steps and progress, maintaining a responsive UI

**Implementation Details:**
- Added `run_async_with_events()` method to `AgentTool` that yields events from sub-agent execution
- Created `handle_function_calls_async_with_agent_tool_streaming()` helper function in `functions.py` to handle AgentTool event streaming
- Modified `_postprocess_handle_function_calls_async()` in `base_llm_flow.py` to use the new streaming mechanism
- Maintains backward compatibility - existing `run_async()` method still works as before


### Testing Plan

_Please describe the tests that you ran to verify your changes. This is required
for all PRs that are not small documentation or typo fixes._

**Unit Tests:**

- [x] I have added or updated unit tests for my change.
- [x] All unit tests pass locally.

**Test Results:**
```
tests/unittests/tools/test_agent_tool_event_streaming.py::test_agent_tool_run_async_with_events_yields_sub_agent_events PASSED
tests/unittests/tools/test_agent_tool_event_streaming.py::test_agent_tool_run_async_with_events_forwards_state_delta PASSED
tests/unittests/tools/test_agent_tool_event_streaming.py::test_agent_tool_event_streaming_in_runner PASSED

======================== 3 passed, 9 warnings in 8.07s =========================
```

**Manual End-to-End (E2E) Tests:**

**Setup:**
```bash
cd contributing/samples/agent_tool_event_streaming
adk web .
```

**Test Steps:**
1. In the web UI, select `agent_tool_event_streaming` from the dropdown
2. Ask: "Research the history of artificial intelligence"
3. Observe the Events panel in real-time

**Expected Behavior:**
- Coordinator agent's function call to `research_agent` appears immediately
- Multiple events from `research_agent` stream progressively (not all at once)
- Events include intermediate model responses showing step-by-step progress
- Final function response event appears after all sub-agent events

**Screenshots:**
<img width="1589" height="788" alt="Screenshot 2025-12-20 at 11 20 12 PM" src="https://github.com/user-attachments/assets/101b969e-292e-41a5-b494-8cdf7b1f5322" />
<img width="1632" height="753" alt="Screenshot 2025-12-20 at 11 20 23 PM" src="https://github.com/user-attachments/assets/31b771de-400c-4209-96b1-92ccf89c2827" />
<img width="1634" height="767" alt="Screenshot 2025-12-20 at 11 21 47 PM" src="https://github.com/user-attachments/assets/59a8e67c-fc2e-46b1-81d8-58c3eb169eb7" />


### Checklist

- [x] I have read the [CONTRIBUTING.md](https://github.com/google/adk-python/blob/main/CONTRIBUTING.md) document.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have manually tested my changes end-to-end.
- [x] Any dependent changes have been merged and published in downstream modules.

### Additional context

This PR will have merge conflicts with PR #3848 (streaming-tools-non-live) as both modify:
- `src/google/adk/flows/llm_flows/functions.py`
- `src/google/adk/flows/llm_flows/base_llm_flow.py`

if we are planning to merge this pr first, then i will have to rebase again and fix the conflicts of #3848

**Note:** These are separate features (PR #3848 handles streaming tools in non-live mode, while this PR handles AgentTool event streaming), but they touch the same code paths.

**Sample Code:**
A complete sample demonstrating the feature is available at:
`contributing/samples/agent_tool_event_streaming/`

**Key Benefits:**
- Enables real-time visibility into sub-agent execution progress
- Improves UX for hierarchical multi-agent systems
- Maintains backward compatibility

sarojrout committed:
feat(agent-tool): Add event streaming propagation from AgentTool sub-agents to Runner

sarojrout committed:
test(agent-tool): Fix state delta forwarding test to verify actual state propagation

sarojrout committed:
refactor(agent-tool): Extract duplicated Runner setup and tool result building logic into helper methods

sarojrout committed:
fixed the path of the samples

ryanaiagent commented:
Hi @sarojrout , Thank you for your contribution! We appreciate you taking the time to submit this pull request. 
Can you please fix the failing unit tests before we can proceed with the review.

sarojrout committed:
fix(telemetry): Wrap async generators with Aclosing for telemetry compatibility

Wrap handle_function_calls_async_with_agent_tool_streaming and agent_tool.run_async_with_events with Aclosing to ensure telemetry test passes.

sarojrout committed:
chore: Re-run CI

sarojrout committed:
perf(functions): Use set for efficient membership testing in agent tool filtering

sarojrout committed:
removed unused imports

sarojrout commented:
/gemini review

sarojrout committed:
formatted

shpodg commented:
@sarojrout It is unexpected result when `Token Streaming`  enabled in adk web. I think the `run_config` should be inherited in subAgent run_async.

sarojrout committed:
fix(agent_tool): Pass run_config to sub-agent runner for Token Streaming support

sarojrout committed:
Merge branch 'main' into feat/agent-tool-event-streaming

shpodg commented:
@sarojrout Found another issue: the execution details of the sub agent, such as the tools used, are exposed to the parent agent when async_run in next turn which is not only agentTool and response.

antolarv commented:
Thank’s for the feature, using the patch right now! Hope this will arrive soon in the official

sarojrout committed:
Merge branch 'main' into feat/agent-tool-event-streaming

sarojrout commented:
> Hi @sarojrout , Thank you for your patience here. I apologize for the delay in getting to this review; I know this has been sitting for a while. This PR has merge conflicts that require changes from your end. Could you please rebase your branch with the latest main branch to address these? Once this is complete, please let us know so we can proceed with the review.

done @ryanaiagent . can you pls review?

sarojrout committed:
fixed the linting error and unit tests

sarojrout committed:
review comments incorporated #3991

sarojrout committed:
review comments incorporated #3991

sarojrout commented:
/gemini review

sarojrout committed:
removed redundant logic #3991

sarojrout committed:
Merge branch 'main' into feat/agent-tool-event-streaming

sarojrout committed:
Merge branch 'main' into feat/agent-tool-event-streaming

mattmcc-attest commented:
Yeah i'd really like this, is there any update? @sasha-gitg

PhilippeMoussalli commented:
any updates on this @sasha-gitg ?

DeanChensj commented:
Thank you for the contribution and the effort you put into enabling real-time visibility for sub-agents!

However, I think it is better not to merge this approach as it goes against ADK's core architectural separation between Tools and Multi-Agent Orchestration:

AgentTool is designed for black-box encapsulation: Its contract as a Tool is to take arguments and return a final result string/dict to the parent model without leaking its internal execution trajectory or intermediate events into the parent's event stream.

For streaming & real-time sub-agent observability, prefer TaskAgent / SingleTurnAgent (and Graph/Workflow coordination): In ADK, multi-agent patterns that require event streaming, state delta propagation, and full trajectory visibility are natively supported as first-class citizens using TaskAgent, SingleTurnAgent, or workflow schedulers rather than wrapping an agent inside a function-calling tool.
Introducing event streaming into AgentTool creates significant complexity in functions.py and blurs the boundary between atomic tool execution and agent orchestration. We encourage migrating use cases that need sub-agent event streaming to use TaskAgent or explicit multi-agent workflows.

sarojrout commented:
> Thank you for the contribution and the effort you put into enabling real-time visibility for sub-agents!
> 
> However, I think it is better not to merge this approach as it goes against ADK's core architectural separation between Tools and Multi-Agent Orchestration:
> 
> AgentTool is designed for black-box encapsulation: Its contract as a Tool is to take arguments and return a final result string/dict to the parent model without leaking its internal execution trajectory or intermediate events into the parent's event stream.
> 
> For streaming & real-time sub-agent observability, prefer TaskAgent / SingleTurnAgent (and Graph/Workflow coordination): In ADK, multi-agent patterns that require event streaming, state delta propagation, and full trajectory visibility are natively supported as first-class citizens using TaskAgent, SingleTurnAgent, or workflow schedulers rather than wrapping an agent inside a function-calling tool. Introducing event streaming into AgentTool creates significant complexity in functions.py and blurs the boundary between atomic tool execution and agent orchestration. We encourage migrating use cases that need sub-agent event streaming to use TaskAgent or explicit multi-agent workflows.

hey @DeanChensj , thanks for your response after a long time! I can understand, now there are many options to handle this scenario. But 8 months back when i raised this PR, that time, there were not many options to think about to handle this kind of use cases and i thought of adding this feature as part of AgentTool(as part of non-deterministic workflow) itself.I would like to thank everyone of you who upvoted for this feature and spent time to go through this PR and understand the usecases!

DeanChensj commented:
Hi @sarojrout, you are absolutely right. When you opened this PR 8 months ago, the ADK didn't have the mature orchestration primitives (like TaskAgent) that we have today, and your approach was a great initiative to solve a very real pain point for the community at that time.

I sincerely apologize for the long delay in reaching this conclusion and reviewing this properly. We deeply appreciate the time, effort, and high-quality code you put into this, as well as everyone in the community who participated and shared their use cases. This discussion was highly valuable for us. Thank you again for your understanding and your great contribution!
